### ISSUE TITLE
ND4J Very slow

### ISSUE THREAD

**Issue by maitri982** [deeplearning4j/nd4j#1570]
We created our own NN using Java.  We implemented threading in Java to speed things up, but really still need more speed of course. So...

We bought a pretty good rig that we assembled to so we could leverage ND4J matrices operations on the GPU.  The machine is Linux (CentOS) with 2 Nvidia Titan X Pascal cards...which have something like 3500 cores each.  We are using CUDA 8.x.

We converted our code to use matrices instead of the traditional Java arrays (we were using Unsafe arrays).  I was expecting a huge uptick in performance.  Instead we saw the opposite.  Our Java threaded code is 2x faster than ND4J!!!  Makes zero sense.  And yes...the GPU is cranking when we run our code on it.

I can provide example code for the FeedForward method the old way and the new way.  BTW...we also tried ot just remove the CUDA so it would thread on the CPU to see what would happen...that was so slow it never got past epoch 1 after 20 minutes so we killed it.  I really hope someone can help with this...thanks!

Here is the code:

```

for (int layerNum = 1; layerNum < NumberOfLayers; layerNum++) {

                MatrixLayer layer  = Layer[layerNum];
                                
                //for each node in this layer
                for (int nodeNum = 0; nodeNum < layer.numberOfNodes; nodeNum++) {
                   
                    //this is the bias being added to the node as a starting point before we add in the weighted inputs
                    float inputTotal = layer.nodeThreshold[nodeNum];

                    //accumulate the inputs of the weights of the input links times their input value
                    for (int nodeInputNum = 0; nodeInputNum < layer.node[0].length; nodeInputNum++) {
                        inputTotal += layer.input[nodeInputNum] * layer.node[nodeNum][nodeInputNum];
                    }

                    //apply sigmoid right here instead of calling a method
                    layer.nodeOutput[nodeNum] = (float)(1/(1+exp(-inputTotal)));
 
                }
                
                // Unless we have reached the last layer, assign the layer's output vector to the next layer's input vector
                if (layerNum != NumberOfLayers-1) {
                    Layer[layerNum].copyOutputVectorTo(Layer[layerNum + 1].input);
                }
            }


New ND4J Code using matrices
//iterate from first hidden layer up to output layer feedingforward the outputs of each layer
            for (int layerNum = 1; layerNum < NumberOfLayers; layerNum++) {

                MatrixLayer layer  = Layer[layerNum];
                
                INDArray biasedWeights = layer.node.mulRowVector(layer.input);
                layer.nodeOutput = biasedWeights.sum(1).transpose();
                layer.nodeOutput.addi(layer.nodeThreshold);
                // Transform sums in place
                Transforms.sigmoid(layer.nodeOutput, false);

                // Unless we have reached the last layer, assign the layer's output vector to the next layer's input vector
                if (layerNum != NumberOfLayers-1) {
                    layer.copyOutputVectorTo(Layer[layerNum + 1].input);
                }
            }
```

**Comment by raver119** [comment]
Show me shape for **layer.input** and **layer.nodeOutput** please.

**Comment by maitri982** [comment]
These are all INDArrays in our MatrixLayer class:

`this.node                    = Nd4j.zeros(numberOfNodes, numberOfInputs);
        this.nodeWeightDiff         = Nd4j.zeros(numberOfNodes, numberOfInputs);
        this.nodeMatrixWorking        = Nd4j.zeros(numberOfNodes, numberOfInputs);
        this.nodeOutput                = Nd4j.zeros(numberOfNodes);
        this.nodeThreshold             = Nd4j.zeros(numberOfNodes);
        this.nodeThresholdDiff         = Nd4j.zeros(numberOfNodes);
        this.signalError             = Nd4j.zeros(numberOfNodes);
        this.signalErrorBiased        = Nd4j.zeros(numberOfNodes);
        this.input                     = Nd4j.zeros(numberOfInputs);`

**Comment by raver119** [comment]
That's great, but still - shapes need, show them please :)

**Comment by maitri982** [comment]
Well...different for each layer.  We are doing MNIST so input layer has numberOfNodes = 784.    The hidden layer we have tried all kinds of settings from small like 5 nodes to 2500 nodes.  This is fully connected from input layer so the numberOfInputs is always 784 for first hidden layer. So shape of hidden layer has been like [5][784] all the way to [2500][784].   It is very slow at all these hidden layer settings.

**Comment by raver119** [comment]
That explains a lot.

You just can't use GPUs in this way, and expect them to be any well-performing. Doesn't matters how cool your GPUs are. That's kinda basics of CUDA programming, as well as general "vectorized" programming.

You might want to look into DL4j, to see how we do that. Basically - batches of inputs are submitted to GPU for processing. So instead of rows - we work on matrices, where each row (or tensor) is "separate" example.

As for timing, you might want to take a look here: http://52.38.112.126/nd4j-remote/
and you might notices difference between similar tests being done on different backends, i.e. Scalars small and Scalars xl. There's the same op behind the scenes, but different length of operands. And you can see yourself - on small test cpu takes 1 us to finish (in reality - 0.7 us), and cuda takes 10 us. And XL operand takes 27 us on cpu, but the same 10 us on CUDA.

That happens due to CUDA kernel launch price - it's flat price around 7000 nanoseconds per kernel ivocation, not counting memory transfers.

**Comment by raver119** [comment]
Btw, one of DL4j tests show on that page, is exactly MNIST using simple MLP - as you can see yourself - 2 milliseconds on CUDA vs 17 milliseconds on CPU.

**Comment by maitri982** [comment]
Thanks raver for your quick and thorough responses.

I will note a few things:

- we are doing online training here.  So 60K samples feed-->calc error-->backprop for each one individually...no batching of training samples.
- in our pure java code, we have outer layer loop and two inner loops to loop through a layers nodes and the weights...the inner loops are what were expensive in terms of processing time in Java

For the matrix code we have eliminated the inner loops altogether.  So we are passing a large 2D matrix that represents the entire layer or nodes and weights to the GPU...is that not batching?  Focusing on just feedforward for the moment, you have to process each layer in sequence for each sample (assuming online learning) so not sure how we could batch any more than we are right now.  We can only pass one layer at a time before proceeding to the next layer in the NN.

Thanks again for your feedback!

**Comment by raver119** [comment]
That brings the same question: show me shape information for arrays, and i'll tell you what's wrong :)

However, general problem is 100% the same i've just explained: GPU shines only when your data is big enough to make busy all MPs of your GPU. You've mentioned Titan X, so that's 28 multiprocessors.

**Comment by raver119** [comment]
Also, to be mentioned: there's many other performance killers waiting for you there: JVM garbage collector mechanic is painful at this moment for us, INDArray operations on different orders (C/F) are expensive, gemm performance on different orders differs, etc etc etc.

**Comment by maitri982** [comment]
Surely seems like the "load" penalty offsets any gain in parallelism for our current approach.   

```
this.node                    = Nd4j.zeros(numberOfNodes, numberOfInputs);
        this.nodeWeightDiff         = Nd4j.zeros(numberOfNodes, numberOfInputs);
        this.nodeMatrixWorking        = Nd4j.zeros(numberOfNodes, numberOfInputs);
        this.nodeOutput                = Nd4j.zeros(numberOfNodes);
        this.nodeThreshold             = Nd4j.zeros(numberOfNodes);
        this.nodeThresholdDiff         = Nd4j.zeros(numberOfNodes);
        this.signalError             = Nd4j.zeros(numberOfNodes);
        this.signalErrorBiased        = Nd4j.zeros(numberOfNodes);
        this.input                     = Nd4j.zeros(numberOfInputs);
```
The shapes are all shown above and vary by layer.  Most are single dimensional arrays that are the number of nodes in the layer.  Exception is input array which is the number of inputs to this layers nodes.  For first hidden layer that's 784.

I hear ya on extras like GC causing pain, etc.  

BTW....we are getting near 99% on MNIST with deep network with all of the following true:

- no dropout
- no convolutions
- no batching...strictly online learning
- sigmoid function
- fully connected layers
- deep network (5 hidden layers)
- no regularization
- no validation set or  as-we-go hyperparameter adjustments

Just need speed for more "knob turning" testing on less time :)

We may implement batching so we are throwing 3D matrices (depth is training samples in batch...row is node and column is weight) at the GPU for feedforward and error calculations.

**Comment by raver119** [comment]
Well, if you're sending examples to training as 784-elements vector, it won't ever load your GPU even to 1%. That's just nothing for Titan X, it spends 99% of time executing micro kernels, that never big enough to pay back. And time spent will be just awful. At other hand as i've said before - with DL4j, single layer MLP trainin on MNIST corpus takes 2 milliseconds per batch, and will be faster soon.

**Comment by agibsonccc** [comment]
I'm closing this, it's fairly obvious it wasn't nd4j. Hand rolling your own neural nets for commercial activities based on nd4j is more or less our wheel house :D. We're happy to help where we can, but it's fairly obvious it was just your implementation. Please come in to the gitter channel if you other specific discussions like this.
